Sources/Core/Sources/ECS/Systems/PlayerInputSystem.swift

```diff
@@ -146,8 +146,8 @@ public final class PlayerInputSystem: System {
         guiState.currentMessageIndex = nil
         return true
       } else if event.key == .delete {
-        if !message.isEmpty {
-          guiState.messageInput?.removeLast()
+        if !message.isEmpty && guiState.messageInputCursor < guiState.messageInput?.count ?? 0 {
+          guiState.messageInput?.remove(at: message.index(message.endIndex, offsetBy: -guiState.messageInputCursor - 1))
```

Comment:
Since the conversion from cursor to index happens quite a few times, I reckon it should be a computed property of `GUIState`.

```swift
// GUIState.swift
// ...
  var messageInputCursorIndex: String.Index {
    return message.index(message.endIndex, offsetBy: -messageInputCursor)
  }
// ...
```

In this one particular spot, the index is offset by -1 from all of the other times this index is calculated, so you can just use `message.index(before: guiState.messageInputCursorIndex)`